xmake/rules/c++/modules/modules_support/builder.lua

```diff
@@ -0,0 +1,407 @@
+
+    local objectfiles = dependency_scanner.sort_modules_by_dependencies(sourcebatch.objectfiles, modules)
+    _build_modules(target, batchcmds, objectfiles, modules, opt)
+end
+
+-- generate headerunits for batchjobs
+function build_headerunits_for_batchjobs(target, batchjobs, sourcebatch, modules, opt)
+
+    local user_headerunits, stl_headerunits = dependency_scanner.get_headerunits(target, sourcebatch, modules)
+    if not user_headerunits and not stl_headerunits then
+       return
+    end
+    -- we need new group(headerunits)
+    -- e.g. group(build_modules) -> group(headerunits)
+    opt.rootjob = batchjobs:group_leave() or opt.rootjob
+    batchjobs:group_enter(target:name() .. "/build_headerunits", {rootjob = opt.rootjob})
+  
+    -- build stl header units first as other headerunits may need them
+    if stl_headerunits then
+        opt.stl_headerunit = true
+        _build_headerunits(target, batchjobs, stl_headerunits, opt)
+    end
+    if user_headerunits then
+        opt.stl_headerunit = false
+        _build_headerunits(target, batchjobs, user_headerunits, opt)
+    end
+end
+
+-- generate headerunits for batchcmds
+function build_headerunits_for_batchcmds(target, batchcmds, sourcebatch, modules, opt)
+
+    local user_headerunits, stl_headerunits = dependency_scanner.get_headerunits(target, sourcebatch, modules)
+    if not user_headerunits and not stl_headerunits then
+       return
+    end
+    -- build stl header units first as other headerunits may need them
+    if stl_headerunits then
+        opt.stl_headerunit = true
+        _build_headerunits(target, batchcmds, stl_headerunits, opt)
+    end
+    if user_headerunits then
+        opt.stl_headerunit = false
+        _build_headerunits(target, batchcmds, user_headerunits, opt)
+    end
+end
+
+-- append headerunits objectfiles to link
+function append_dependency_objectfiles(target)
+
+    local cachekey = target:name() .. "dependency_objectfiles"
+    local cache = compiler_support.localcache():get(cachekey)
+    if cache then
+        if target:is_binary() then
+            target:add("ldflags", cache, {force = true, expand = false})
+        elseif target:is_static() then
+            target:add("arflags", cache, {force = true, expand = false})
+        elseif target:is_shared() then
+            target:add("shflags", cache, {force = true, expand = false})
+        end
+    end
+end
+
+-- get or create a target module mapper
+function get_target_module_mapper(target)
+
+    opt = opt or {}
+    local memcache = compiler_support.memcache()
+    local mapper = memcache:get2(target:name(), "module_mapper")
+    if not mapper then
+        mapper = compiler_support.localcache():get(_target_module_map_cachekey(target)) or {}
+
+        memcache:set2(target:name(), "module_mapper", mapper)
+    end
+
+    return mapper
+end
+
+-- get a module or headerunit from target mapper
+function get_from_target_mapper(target, name)
+
```

Comment:
remove space line

Reply:
just to be sure, when should i add / doesn't add a blank line at the begin of the function ? (because i saw it on some others function)

Reply:
we need not add it if be small function body.